### DISCUSSION TITLE
add "optional" flag to plugin configuration

### DISCUSSION THREAD

**Discussion by janschab** [nrwl/nx#23140] [Feature Requests]
Currently, to build the application, the presence of packages for plugins defined in nx.json > plugins is required. Not all plugins are necessary for building the project, such as eslint or jest. When I omit devDependencies (among which are packages of these plugins) during installation, nx cannot build the graph. I don't want to add these packages as regular dependencies because they are not needed to build the application.

I propose adding an 'optional' flag to the plugin configuration so that nx can continue building without throwing an error in case packages for such marked plugins are missing.

**Comment by janschab** [comment]
Added a sample solution that introduces the 'optional' flag to plugin configuration in nx. Feel free to test it out and share your thoughts! 
[commit](https://github.com/nrwl/nx/commit/803c29794c0ec661a5a50f0972ad9afb85e540f8)

**Comment by jaysoo** [comment]
Hi @janschab Thanks for the feature suggestion. I don't think `devDependencies` and optional is the right solution, as the partial graph from the subset of plugins may not be correct. It sounds like you are optimizing install times, is this for CI?

**Reply by janschab** [reply]
I realize that my idea may not be right. I optimize the process of building packages for testing and production environments. It would be most beneficial if I could install the necessary packages and run the build script without needing the graph. I have a Standalone Repository with a root buildable application. Do you have any ideas on how to bypass building the graph and only build one thing? Or any other hint. I know it sounds like going against the use of Nx

**Reply by jaysoo** [reply]
It's not a use-case we support (to run tasks without Nx graph). Running script without Nx only works if you are not using any Nx executors. For example, if you have `"build": "next build"` if your `package.json`, and Nx is just invoking that script when you run `nx build`. However, doing so forgoes the benefit of using Nx, so would not recommend it.

For CI pipelines, you can cache the `node_modules` folder, which should speed up install times. For GitHub actions, there is a [guide here](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows).

**Reply by janschab** [reply]
Thank you @jaysoo for your responses and tips. I appreciate it.
